transformers/generate/sampler.py

```diff
@@ -0,0 +1,210 @@
+from collections import namedtuple
+import warnings
+
+import torch
+import torch.nn.functional as F
+from tqdm import trange
+
+SamplerConfig = namedtuple("SamplerConfig", ["temperature", "k", "p", "repetition_penalty"])
+
+
+def new_sampler(
+    model, temperature=1.0, k=0, p=0, repetition_penalty=1.0, device=torch.device("cpu")
+):
+    """ Factory function that returns the appropriate sampler with regards
+    to the model passed as a parameter.
+
+    Only single stacks are currently supported.
+    """
+
+    sampler_config = SamplerConfig(
+        temperature=temperature, k=k, p=p, repetition_penalty=repetition_penalty
+    )
+
+    has_encoder = callable(getattr(model, "encode", None))
```

Comment:
why don't we just check if the model is an instance of `PreTrainedEncoderDecoder` here?

Reply:
Yes, that's good idea.